### PR TITLE
[FLINK-25817] Use working directory to persist local state

### PR DESCRIPTION
**Pull request by tillrohrmann**
This PR implements [FLIP-201](https://cwiki.apache.org/confluence/display/FLINK/FLIP-201%3A+Persist+local+state+in+working+directory). The goal is to use the working directory to store local state information so that a process can recover from process/node failures as long as the working directory is available after the restart.

The important commits are:

68200f5caa6187f267b288925fdc5bae811e0a8b: This commit lets the `TaskLocalStateStoreImpl` persist the `TaskStateSnapshot` to the working directory. This is required to recover any local state after a process failure.

ed11d5fc78748dc48db18a608dc08062be9dfd4c and 65479221d6cfde6ce89f79f53cbbe0e2b3f26dae: These two commits let the `TaskExecutor` persist the current slot allocations in the working directory. Upon restart of a `TaskExecutor`, it will read this information and recover the last slot allocation state. That way we can couple the clean up of local state to the slot allocation because the local state and slot allocation can both be recovered from the working directory.

This PR is based on #18542.

### CODE REVIEW

File: docs/layouts/shortcodes/generated/cluster_configuration.html
Comment (dmvk) [inline docs/layouts/shortcodes/generated/cluster_configuration.html:99]: This is mapping multiple directories into a single directory. Can you add an explanation on which directory would be picked in case of multiple io dirs?

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/EmbeddedRocksDBStateBackend.java
Comment (tillrohrmann) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/EmbeddedRocksDBStateBackend.java:324]: Users can still configure multiple RocksDB working directories via `state.backend.rocksdb.localdir`. Hence, I think that this should be ok.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/TaskManagerRuntimeInfo.java
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/TaskManagerRuntimeInfo.java:68]: I'll add it in the corresponding PR that introduces the change.

File: flink-runtime/src/test/java/org/apache/flink/runtime/util/JvmExitOnFatalErrorTest.java
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/util/JvmExitOnFatalErrorTest.java:193]: It won't hurt. I'll add it in the corresponding PR.

File: flink-runtime/src/test/java/org/apache/flink/runtime/util/BlobServerResource.java
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/util/BlobServerResource.java:18]: Getting rid of the `BlobServerResource`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManager.java
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManager.java:296]: The `Reference` is not intended to give guarantees that something is no longer in use. The idea is to convey the notion whether you own it or not and, thus, are responsible for the clean up. The component handing out these references is responsible that all components are shut down before cleaning up the value of the reference.

At the moment I think it is fine as is. If we need ref counting in the future, we can add it later.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/FileSlotAllocationSnapshotPersistenceService.java
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/FileSlotAllocationSnapshotPersistenceService.java:86]: Yes, I will remove it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/FileSlotAllocationSnapshotPersistenceService.java
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/FileSlotAllocationSnapshotPersistenceService.java:105]: Yes, this is probably better.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/FileSlotAllocationSnapshotPersistenceService.java
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/FileSlotAllocationSnapshotPersistenceService.java:42]: We probably can. The biggest change will be to add the infrastructure to add and retrieve the correct version and the corresponding serializer/deserializer. Since this is something we can easily add as a follow up I would like to tackle this later.

The only limitation I can see by not doing it right now is that we don't support version upgrades while keeping the same working directory. I think this is fine for the first version.

File: docs/layouts/shortcodes/generated/cluster_configuration.html
Comment (tillrohrmann) [reply docs/layouts/shortcodes/generated/cluster_configuration.html:99]: Yes, will add it in the corresponding PR for this change.

**Comment by tillrohrmann** [comment]
The PR is now ready for review. Thanks a lot for your review so far @dmvk.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorRecoveryTest.java
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorRecoveryTest.java:43]: Should use junit5

File: flink-tests/src/test/java/org/apache/flink/test/recovery/LocalRecoveryITCase.java
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/test/recovery/LocalRecoveryITCase.java:75]: FYI: in junit test classes and methods can be package private.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManagerTest.java
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManagerTest.java:39]: Shouldn't use the curator import

**Review by zentol — COMMENTED** [review]
I did a quick pass and it looks overall fine. A rebase would be good, and I'd take a closer look on Monday.

What I'm currently wondering is how the JM is notified that these slots are still allocated.
The RM knows it from the first slot report (and thus will not attempt re-allocate the slots).

But we aren't sending slots reports to the JM, nor are we offering the slots as this only happens when instructed to do so by the RM, which shouldn't do that since it considers them to be already allocated.

**Comment by tillrohrmann** [comment]
Thanks for the first review @zentol. I will do a rebase right away.

Concerning your question about how the `TaskExecutor` offers its slots to a `JobMaster` upon start. It works by loading the persisted allocation snapshot. If this happens for a slot that is allocated for a job with `jobId`, then a new `Job` will registered at the `jobTable`. If this happens, then we will add the job to the `jobLeaderService` that will look up the leading `JobMaster`. If a leading `JobMaster` is found, then the `TaskExecutor` will try to connect to it and offer the slots it has reserved for this job.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java:193]: Will fix it.

File: flink-tests/src/test/java/org/apache/flink/test/recovery/LocalRecoveryITCase.java
Comment (tillrohrmann) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/LocalRecoveryITCase.java:75]: Will change it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorRecoveryTest.java
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorRecoveryTest.java:43]: Will change it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManagerTest.java
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManagerTest.java:39]: Indeed. will change it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: I will see whether I can unify both methods. I think this is a good improvement.

**Comment by zentol** [comment]
What happens if the new TM offers its slots before the loss of the old TM is noticed?

The slot pool will treat the offer slots as duplicate registrations (since the AllocationIDs match) and ignore them, but that means the taskManagerLocation will be incorrect (and thus they will be unusable).

If the job then notices the TM loss and fails the job, all the slots belonging to the old TM will be freed.
At that point, the JM lacks slots, the TM thinks they are allocated to the job (after all the offer was accepted), as does the RM (slot reports from TM).
But the JM is just waiting for slots to arrive?

**Comment by zentol** [comment]
I actually need to check what the RM would do in this scenario...

**Comment by tillrohrmann** [comment]
> What happens if the new TM offers its slots before the loss of the old TM is noticed?
> 
> The slot pool will treat the offer slots as duplicate registrations (since the AllocationIDs match) and ignore them, but that means the taskManagerLocation will be incorrect (and thus they will be unusable).
> 
> If the job then notices the TM loss and fails the job, all the slots belonging to the old TM will be freed. At that point, the JM lacks slots, the TM thinks they are allocated to the job (after all the offer was accepted), as does the RM (slot reports from TM). But the JM is just waiting for slots to arrive?

The new TM will first register at the JM before offering the slots. This will deregister any old TM registrations under the same `ResourceID`.

**Comment by zentol** [comment]
huh. Quite a surprise that we already handled that case gracefully.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/FileSlotAllocationSnapshotPersistenceService.java
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/FileSlotAllocationSnapshotPersistenceService.java:123]: Something to keep in mind is that this allows someone with write access to the workingDirectory to execute arbitrary code.
I think we'd want to migrate this to something like jackson soon (i.e., before the release), which should be pretty straight-forward.

**Review by zentol — APPROVED** [review]

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/FileSlotAllocationSnapshotPersistenceService.java
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/FileSlotAllocationSnapshotPersistenceService.java:123]: This is a good point. I will try to open a follow-up PR for it asap.

**Comment by tillrohrmann** [comment]
Thanks a lot for the review @zentol. I've rebased the PR. I will merge it once AZP gives green light.

**Pull request closed without merging**
